### DISCUSSION TITLE
[General] 502 Server Error when calling balanceOf of the contract

### DISCUSSION THREAD

**Discussion by mwrrlvkixj** [zkSync-Community-Hub/zksync-developers#118] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

I tried to call balanceOf of the contract to get the balance for a specific address

### What is the specific issue or error you're encountering?

I tried to call balanceOf of the contract and got this error:
```
502 Server Error: Bad Gateway for url: https://mainnet.era.zksync.io/
```

### Can you share the error messages or logs you're receiving, if any?

```
502 Server Error: Bad Gateway for url: https://mainnet.era.zksync.io/
```

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

python web3

### Can you provide the relevant portions of your contract code where the issue is occurring?

```
value = w3_contract.functions.balanceOf(
                safe_to_check_sum_address(address)
            ).call(block_identifier=block_number)
```

### Have you tried to isolate the problem, and if so, what were the results?

Not sure how  to do that

### What steps have you already taken to try to resolve the issue?

Nothing can be done from our side

**Comment by githubdoramon** [comment]
@mwrrlvkixj can you provide the address of this contract?

**Reply by mwrrlvkixj** [reply]
Other methods also experience it:
```
zks_getBlockDetails 12644481, 0 Exception: 502 Server Error: Bad Gateway
```

**Reply by MexicanAce** [reply]
@mwrrlvkixj Can you attempt to hit the API directly with `curl`? The below command is successful from my local machine.

```bash
curl https://mainnet.era.zksync.io/ \
  -X POST \
  -H "Content-Type: application/json" \
  --data '{"method":"zks_getBlockDetails","params":[12644481],"id":1,"jsonrpc":"2.0"}'
```

**Reply by mwrrlvkixj** [reply]
I couldn't reproduce it either, this error happened very randomly... maybe cuz i hit the api too frequently and your gateway was not able to handle?

**Reply by MexicanAce** [reply]
These API endpoints have rate-limiting to help prevent DDOS attacks, and this might be what you've experienced and started receiving `502 Bad Gateway` errors when that limit was reached. How often are you receiving these errors?

**Reply by mwrrlvkixj** [reply]
I'm not sure about that:
1. i reduced the number of calls per sec, still saw this error.
2. rate limit should be 429 rather than 502

**Reply by MexicanAce** [reply, marked as answer]
Correct, I would've also expected a 429. This could be caused from a possible, temporary outage with the node during that timeframe.

As this issue is not reproducible by either or us anymore, I'm gonna chalk this up to a temporary outage. If it happens again, please feel free to update this Discussion. I'll keep an eye on it 👀 .

**Comment by springzh** [comment]
We also encountered this issue https://github.com/zkSync-Community-Hub/zkync-developers/discussions/119 . Message of "execution reverted"

**Comment by grantimagen** [comment]
It seems like the issue is related to a server error (502 Bad Gateway) when trying to interact with the zkSync mainnet API. This could be due to temporary server issues or maintenance on zkSync's end. You can try again later or check zkSync's status page for any updates on the issue.

**Comment by cahillnick** [comment]
It seems that there is an issue with the connection to the zkSync mainnet API.
 You may want to retry the operation later or check for any updates from zkSync regarding their API status.

**Comment by shelleyalexander** [comment]
The error "502 Server Error: Bad Gateway" typically indicates an issue with the server or network connectivity. This error is likely not related to your contract code or the specific function you're trying to call. It's possible that there is a temporary issue with the zkSync server. You may want to try again later or contact zkSync support for assistance.

**Comment by marcellusjustin** [comment]
The user is encountering a "502 Server Error: Bad Gateway" when trying to call the balanceOf function of a contract on the ZkSync mainnet using the Python web3 library. They have not made any recent changes to the contract and have not tried to isolate the problem. They have also not taken any steps to resolve the issue as they believe it is outside of their control.

**Comment by mooralfeapur** [comment]
The user is encountering a 502 Server Error: Bad Gateway for url: https://mainnet.era.zksync.io/ when trying to call balanceOf of the contract to get the balance for a specific address using python web3 on the Mainnet. The user has not made any recent changes to the contract before encountering this issue. The contract interacts with the web3 external library. The user has provided relevant portions of their contract code where the issue is occurring. The user has not tried to isolate the problem. The user has not taken any steps to try to resolve the issue, as they believe nothing can be done from their side. No repo link is provided.
